User refactor + remove fastapi_users dependency + update fastapi

ikreymer opened the pull request:
Fixes #1050 

Major refactor of the user/auth system to remove fastapi_users dependency. Refactors users.py to be standalone
and adds new auth.py module for handling auth. UserManager now works similar to other ops classes.
Also includes:
- allows updating fastapi to latest
- add webhook docs to openapi (follow up to #1041)

Branch based off #1272 work, and may supersede it fully.

The auth should be fully backwards compatible with fastapi_users auth, including accepting previous JWT tokens w/o having to re-login. The User data model in mongodb is also unchanged. 

API changes:
- Removing the`GET, PATCH, DELETE /users/<id>` endpoints, which were not in used before, as users are scoped to orgs. For deletion, probably auto-delete when user is removed from last org (to be implemented). 
- Rename `/users/me-with-orgs` is renamed to just `/users/me/`

tw4l committed:
Modify PATCH /users/me to require current password

Implements a forked custom user router to make the change, since
overriding fastapi endpoints is not straightforward.

tw4l committed:
Fix linting

tw4l committed:
Another try at fixing linting issues

tw4l committed:
Aaand fix linting for real

tw4l committed:
Instruct mypy to ignore type for user_manager.update

tw4l committed:
Ignore long lines

tw4l committed:
Update with new endpoint

SuaYoo committed:
remove auth check

tw4l committed:
Fixups

tw4l committed:
Add mypy ignore

tw4l committed:
Require user email in /me/password-change body

tw4l committed:
Update password tests

tw4l committed:
Update test_users:test_reset_password_invalid_current

tw4l committed:
Try verifying plaintext against user's stored hash

ikreymer committed:
backend: remove unsafe PATCH routes, remove router override
update model to contain 'password' and 'newPassword'
frontend: match model

ikreymer committed:
rename /users/me-with-orgs to /users/me
update tests

ikreymer committed:
remove unused model

ikreymer committed:
remove get_users_router() altogether

ikreymer committed:
tests: use latest crawler, lower resource usage

ikreymer committed:
remove fastap_users register, reset, verify routers, replace with local routes

ikreymer committed:
remove fastapi_users auth, replace with fastapi OAuth2 jwt token creation + decoding
move auth functionality to auth.py

ikreymer committed:
unify password auth functions

ikreymer committed:
full removal first pass
- move relevant functions from fastapi_user manager to users
- update modules
- remove fastapi_users, update to latest fastapi!
- still more cleanup needed, simplification where possible

ikreymer committed:
update reqs

ikreymer committed:
add missing file

ikreymer committed:
fix req

ikreymer committed:
additional type fixes for latest pydantic, update reqs

ikreymer committed:
refactoring work:
- keep pydantic at 1.x for now (made some model changes that are forward compatible with 2.0)
- ensure users uses _id not id similar to all other objects (will need migration)
- remove userdb, move db functions into users
- remove generic _update, add update_verified, update_password, update_email
- webhooks: add webhooks for openapi!

ikreymer committed:
additional type fixes:
- remove references to 'user_db'
- unify 'get_id' / 'get_user_by_id' to just 'get_by_id'
- fix 'id' -> '_id' reference

ikreymer committed:
add update_invites

ikreymer committed:
change password cleanup, rename UserUpdatePassword -> UserUpdate

ikreymer committed:
more type fixes use UserDB instead of User
rename create->register
rename create_base->_create

ikreymer committed:
merge on_after_register with _create, update
invite + new org logic

ikreymer committed:
remove remaining on_after* methods
consoldiate UserDB->User

ikreymer committed:
more cleanup

ikreymer committed:
update test

ikreymer committed:
test fixes

ikreymer committed:
cleanup, update response models for auth

ikreymer committed:
api function name tweaks

ikreymer committed:
return UserOut model from register also
remove is_active from model, check for 'id' on frontend instead of unused 'is_active'

ikreymer committed:
cleanup

ikreymer committed:
cleanup:
- error code pass: use more consistent lowercase errors, remove old fastapi_users error code
- use 'user_already_exists' and 'invalid_password' for dupe user and invalid password errors
- update frontend error codes checking
- rename self.collection -> self.users in UserManager

ikreymer committed:
users model: don't switch to '_id' for id, key 'id' key for backwards compatibility

ikreymer committed:
auth: add audience to jwt tokens, accept audience from fastapi_users for backwards compatibility

ikreymer committed:
fix OrgOut serialization, use 'id' instead of '_id'

ikreymer committed:
fix orgout model

tw4l commented on backend/btrixcloud/operator.py:
```diff
@@ -211,6 +211,9 @@ class CrawlStatus(BaseModel):
     podStatus: Optional[DefaultDict[str, PodInfo]] = defaultdict(
         lambda: PodInfo()  # pylint: disable=unnecessary-lambda
     )
+    # podStatus: Optional[
+    #    DefaultDict[str, Annotated[PodInfo, Field(default_factory=PodInfo)]]
+    # ]
```
Curious about why this is commented out here, does it not work as actual type checking?

ikreymer replied:
Oh, this was part of migration to pydantic 2.0, which decided not do at this point! That can be a separate PR, for now, sticking with pydantic 1.x

ikreymer replied:
This change will be needed for pydantic 2.0 migration, maybe can add a comment and leave it in for now. The current type declaration does not work with pydantic 1.x (other changes were backwards compatible).

ikreymer replied:
Added comment.

tw4l commented on backend/btrixcloud/orgs.py:
```diff
@@ -601,7 +602,7 @@ async def accept_invite(token: str, user: User = Depends(user_dep)):
 
         await ops.add_user_by_invite(invite, user)
-        await user_manager.user_db.update(user)
+        # await user_manager.update(user)
```
Is this not necessary? Might be better to remove rather than comment out if so

ikreymer replied:
Removed the general `user_manager.update` so commented out to return to later. Looks like what's actually updated (in `accept_user_invite`) is invites so should call `user_manager.update_invites`, will update to that. Nice catch!

ikreymer replied:
Moved `user_manager.update_invites` into `invites.accept_user_invite` to be more consistent / update immediately after removing.

tw4l commented on backend/btrixcloud/users.py:
```diff
@@ -69,96 +67,150 @@ def set_org_ops(self, ops):
+        user = await self.get_by_email(email)
+        if not user:
+            # Run the hasher to mitigate timing attack
+            # Inspired from Django: https://code.djangoproject.com/ticket/20760
+            get_password_hash(password)
```
Not sure if this was you or from fastapi-users, but nice!

ikreymer replied:
This was copied from fastapi_users, which copied it from Django, figured it was good to keep!

tw4l commented on backend/btrixcloud/users.py:
```diff
@@ -197,276 +247,426 @@ async def create_non_super_user(
+    async def change_email_name(
+        self, user_update: UserUpdateEmailName, user: User
+    ) -> None:
+        """Change password after checking existing password"""
```
Maybe needs a new docstring?

tw4l commented on backend/btrixcloud/users.py:
```diff
@@ -197,276 +247,426 @@ async def create_non_super_user(
+        )
+
+    async def update_invites(self, user: User) -> None:
+        """Update verified status for user"""
```
Another docstring that needs an update

ikreymer replied:
Fixed!

tw4l commented on backend/btrixcloud/users.py:
```diff
@@ -197,276 +247,426 @@ async def create_non_super_user(
+        await self.validate_password(new_password)
+        hashed_password = get_password_hash(new_password)
+        if hashed_password == user.hashed_password:
+            return False
```
I don't think this will always work as expected, as we can get a different hash each time. Probably want to use `auth.verify_password` instead

ikreymer replied:
This wasn't used for verification, but just to check if hash been updated, but, it always will be so doesn't make sense.
Update logic to use `check_password()` for superuser password checking and made this method `_update_password()` to indicate internal user

tw4l commented on backend/btrixcloud/webhooks.py:
```diff
@@ -404,41 +404,35 @@ async def retry_notification(
     return ops
+
+
+def init_openapi_webhooks(app):
+    """add webhooks declarations for openapi"""
+
+    # pylint: disable=unused-argument
+    @app.webhooks.post(WebhookEventType.CRAWL_STARTED)
+    def crawl_started(body: CrawlStartedBody):
+        """Sent when a crawl is started"""
+
+    @app.webhooks.post(WebhookEventType.CRAWL_FINISHED)
+    def crawl_finished(body: CrawlFinishedBody):
+        """Sent when a crawl if finished"""
+
+    @app.webhooks.post(WebhookEventType.UPLOAD_FINISHED)
+    def upload_finished(body: UploadFinishedBody):
+        """Sent when an upload has finished"""
+
+    @app.webhooks.post(WebhookEventType.ADDED_TO_COLLECTION)
+    def added_to_collection(body: CollectionItemAddedBody):
+        """Sent when an archived item (crawl or upload)
+        is added to a collection"""
+
+    @app.webhooks.post(WebhookEventType.REMOVED_FROM_COLLECTION)
+    def remove_from_collection(body: CrawlStartedBody):
+        """Sent when an archived item (crawl or upload)
+        is removed from a collection"""
```
Woohoo!

ikreymer replied:
💯

tw4l reviewed (commented):
Initial comments, testing locally now

tw4l commented:
Manually tested:
- Changing user password via forgot password link
- Chaning user name, email, and password via account settings
- Changing user roles in org via org settings
- Log in and log out
- Usernames set correctly on crawls and workflows after creating and deleting crawls

Aside from the comment about using `auth.verify_password()`, this is looking good to me!

tw4l commented:
I am seeing an error trying to render the API docs:

<img width="451" alt="Screen Shot 2023-10-17 at 4 21 12 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/d0c552c3-5a90-494d-ae5d-48f1c002062e">

Console error:

<img width="1211" alt="Screen Shot 2023-10-17 at 4 21 24 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/ce913492-49a4-4a94-a3c3-3290b800b303">

Update: seeing this as an issue on Chromium Version 92.0.4515.131 (Official Build, ungoogled-chromium) (arm64) on macOS. In Chrome Version 118.0.5993.70 (Official Build) (arm64), Firefox Version 118.0.5993.70 (Official Build) (arm64), and Safari Version 16.5 (17615.2.9.11.6, 17615) it works fine.

Might just be due to the older Chromium version? Going to say this is definitely not a blocker given it's only happening in one older browser and not any of the up-to-date mainstream browsers.

ikreymer committed:
fixes from code review:
- fix doc strings
- add user.update_invites() when removing an invite
- fix logic for printing when superuser password has changed

ikreymer committed:
Merge branch 'main' into remove_fastapi_users

ikreymer committed:
add 'slug' to UserOrgInfoOut

ikreymer committed:
readd default for pop
